### PR TITLE
Refactor: Replace SFINAE by `if constexpr` for `create_mirror*` functions

### PR DESCRIPTION
**Pull request by pzehner**
This PR is a refactor without side effects.

The idea is to simplify the code of the `create_mirror*` functions by applying techniques of the C++ 17 standard. A refactor with the similar approach was drafted by @masterleinad in https://github.com/kokkos/kokkos/pull/5533 about two years ago, but never merged.

The multiple overloaded functions of `Impl::create_mirror`, `Impl::create_mirror_view`, and `Impl::create_mirror_view_and_copy` that use SFINAE are replaced by one function that uses `if constexpr` instead (for Kokkos views, offset views, dynamic views and dynamic rank views). Eventually, the return type of these functions is replaced by `auto`, and decided depending on which branch of the `if constexpr` is used.

The logic of the public `create_mirror*` functions (which is sometimes redundant with the logic of the private functions) is not altered in this PR and will be assessed in an upcoming one.

This PR is part of a of a larger effort which aims to assess https://github.com/kokkos/kokkos/issues/6842.

We successfully built and tested the refactored code for the following configurations:

| Compiler      | Version  | *Min. ver.* | Backend | Version | *Min. ver.* |
|---------------|----------|-------------|---------|---------|-------------|
| Clang         | 14.0.0   | *8.0.0*     | Serial  |         |             |
| GCC           | 8.4.0    | *8.2.0*     | Serial  |         |             |
| GCC           | 8.4.0    | *8.2.0*     | OpenMP  |         |             |
| GCC           | 8.4.0    | *8.2.0*     | CUDA    | 11.4    | *11.0*      |
| Intel Classic | 2023.2.1 | *19.0.5*    | OpenMP  |         |             |
| Intel LLVM    | 2023.2.1 | *2021.1.1*  | OpenMP  |         |             |
| Intel LLVM    | 2024.0.2 | *2022.0.0*  | SYCL    |         |             |
| HIP           | 5.2.0    | *5.2.0*     | ROCM    |         |             |

where "Compiler" is the used compiler, next "Version" is the version of the compiler, next "Min. Vers." is the minimum version of the compiler for Kokkos 4.3.0 according to the documentation, "Backend" is the name of the backend, next "Version" is the version for the backend if necessary, and next "Min. Vers." is the minimum version of the backend supported.

Unfortunately, we could not test any earlier version of the compilers, nor of the backends, on our test machines. Also, please note that some configurations could not be used:

- Clang with OpenMP or CUDA;
- MSVC; and
- NVHPC.

Finally, the SYCL configuration was built, but not tested.

### CODE REVIEW

File: containers/src/Kokkos_DynRankView.hpp [hunk 1]
```diff
@@ -2058,71 +2044,39 @@ inline auto create_mirror(
...
-
-template <class T, class... P, class... ViewCtorArgs,
-          class = std::enable_if_t<
-              Impl::ViewCtorProp<ViewCtorArgs...>::has_memory_space>>
-inline std::enable_if_t<
-    Kokkos::is_space<
-        typename Impl::ViewCtorProp<ViewCtorArgs...>::memory_space>::value &&
+            typename DynRankView<T, P...>::HostMirror::data_type>::value
+        ) {
+      return src;
```
Comment (masterleinad) [inline containers/src/Kokkos_DynRankView.hpp:2063]: Is this the same type as `typename DynRankView<T, P...>::HostMirror`? If we don't intend to change behavior, we should be explicit about keeping the same return types.

File: containers/src/Kokkos_DynRankView.hpp [hunk 2]
```diff
@@ -2058,71 +2044,39 @@ inline auto create_mirror(
...
-        typename Impl::ViewCtorProp<ViewCtorArgs...>::memory_space, T,
-        P...>::view_type>
-create_mirror_view(const Kokkos::DynRankView<T, P...>& src,
-                   const typename Impl::ViewCtorProp<ViewCtorArgs...>&) {
-  return src;
+            P...>::is_same_memspace
+        ) {
+      return src;
+    } else {
+      return Kokkos::Impl::create_mirror(src, arg_prop);
```
Comment (masterleinad) [inline containers/src/Kokkos_DynRankView.hpp:2075]:
```suggestion
      return typename Impl::MirrorDRViewType<
        typename Impl::ViewCtorProp<ViewCtorArgs...>::memory_space, T,
        P...>::view_type{Kokkos::Impl::create_mirror(src, arg_prop)};
```

File: containers/src/Kokkos_DynRankView.hpp [hunk 3]
```diff
@@ -1933,15 +1933,10 @@ struct MirrorDRVType {
...
-    const DynRankView<T, P...>& src,
-    const Impl::ViewCtorProp<ViewCtorArgs...>& arg_prop,
-    std::enable_if_t<!Impl::ViewCtorProp<ViewCtorArgs...>::has_memory_space>* =
-        nullptr) {
-  using src_type = DynRankView<T, P...>;
-  using dst_type = typename src_type::HostMirror;
 
+// collection of static asserts for create_mirror
+template <class... ViewCtorArgs>
+void check_view_ctor_args_create_mirror_drv() {
```
[the remark this answers is not in the document]
Comment (pzehner) [reply containers/src/Kokkos_DynRankView.hpp]: We re-used the `check_view_ctor_args*` functions from `core`.

**Comment by pzehner** [comment]
The code style was fixed using `clang-format`.

**Comment by pzehner** [comment]
Merging done. For merging, please squash the commits, as the history of this branch is quite messy.

**Comment by masterleinad** [comment]
> Merging done. For merging, can you squash the commits, as the history of this branch is quite messy?

Yes, we normally squash commits before merging.

**Comment by pzehner** [comment]
> we should test with Trilinos/Sacado/Stokhos just to make sure this doesn't break their customizations.

Could you recommend any build script for Trilinos, in order to test it?

**Comment by crtrott** [comment]
Retest this please!

**Comment by pzehner** [comment]
@crtrott Could you validate this PR? It seems the failing jobs are due to external factors.

**Review by crtrott — APPROVED** [review]

**Pull request merged**